### PR TITLE
fix: Workspace store quickly reflects add/remove [WEB-635]

### PR DESCRIPTION
**Pull request by mapmeld**
## Description

For the list of pinned workspaces to update immediately after creating or deleting a workspace, we need to update the local workspaces store after create/delete API calls succeed.

In the current PR, the true API calls move to the store `stores/workspaces`, and the modal's line `await createWorkspace(body);` now directs the store to handle the API call and updating itself.

Note: Pinned workspaces are ordered by however the API responds (currently old -> new). For example if I wrote `[w].concat(ws)` it would temporarily add the new workspace at the front of the pinned workspaces array. I rewrote it for consistency then added a `sort()` to the pinned workspaces view to enforce a consistent order.

## Test Plan

- In the left sidebar, scroll down to the last pinned workspace (new workspaces add to the bottom of the list)
- Click the nearby [+] button to open the Create Workspace Modal. Give a name such as "test"
- After clicking OK, the modal disappears and workspace should get simultaneously added on the list.
- Scroll the pinned workspaces again so the new workspace is visible. Delete it and it should simultaneously be removed from the pinned workspaces list

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: webui/react/src/components/NavigationSideBar.tsx
Comment (keita-determined) [inline webui/react/src/components/NavigationSideBar.tsx:293]: idk which one is better but sort by name is more user friendly imo

File: webui/react/src/hooks/useModal/Workspace/useModalWorkspaceCreate.tsx
Comment (keita-determined) [inline webui/react/src/hooks/useModal/Workspace/useModalWorkspaceCreate.tsx:256]: do we need to consider the case if `response` is `undefined`? or we can ignore the case?

File: webui/react/src/components/NavigationSideBar.tsx
Comment (keita-determined) [inline webui/react/src/components/NavigationSideBar.tsx:292]: prob its not from this PR, but spinner position is strange
https://user-images.githubusercontent.com/109113616/207962713-6c2db99c-2e74-4b94-bb65-6b1f3ed7a3a2.mov

**Review by keita-determined — COMMENTED** [review]
the functionalities work good
some comments here

**Comment by keita-determined** [comment]
also, we dont need to move back to workspace list page if deleting other workspace. 

https://user-images.githubusercontent.com/109113616/207963617-1f27d4b2-ccb4-4578-948b-88eb7b110af3.mov

File: webui/react/src/components/NavigationSideBar.tsx
Comment (mapmeld) [reply webui/react/src/components/NavigationSideBar.tsx:293]: Agreed, I was surprised to see that we currently have it oldest -> newest. I will ping design

File: webui/react/src/hooks/useModal/Workspace/useModalWorkspaceCreate.tsx
Comment (mapmeld) [reply webui/react/src/hooks/useModal/Workspace/useModalWorkspaceCreate.tsx:256]: If I format it like this, `createWorkspace` could always return an `undefined`
I don't think that our system supports `createWorkspace` throwing/returning an error, I think I need to call handleError inside the store code.

File: webui/react/src/components/NavigationSideBar.tsx
Comment (mapmeld) [reply webui/react/src/components/NavigationSideBar.tsx:292]: I agree when I was working on pinned workspaces I thought it should be indented a bit more, but I don't know what style options we have then

**Comment by mapmeld** [comment]
> nit: also, we dont need to move back to workspace list page if deleting other workspace.

Let me know if this latest commit makes sense (passing `returnIndexOnDelete = false` to WorkspaceActionDropdown -> useModalWorkspaceDelete) so it doesn't do the redirect if enacted from the sidebar pins

**Review by gt2345 — APPROVED** [review]
LGTM

**Comment by keita-determined** [comment]
when workspace name already exists, it should show error notification. 
i think it's a regression 

https://user-images.githubusercontent.com/109113616/207984029-b570be1b-a929-475c-997e-6c1dce34406a.mov

**Comment by keita-determined** [comment]
> > nit: also, we dont need to move back to workspace list page if deleting other workspace.
> 
> Let me know if this latest commit makes sense (passing `returnIndexOnDelete = false` to WorkspaceActionDropdown -> useModalWorkspaceDelete) so it doesn't do the redirect if enacted from the sidebar pins

i think it does make sense

File: webui/react/src/components/NavigationSideBar.tsx
Comment (mapmeld) [reply webui/react/src/components/NavigationSideBar.tsx]: After convo with design, changed this to sort by `pinned.createdAt` with newest pins at the bottom (luckily this was always in our DB and just needed to be sent by the API and can be sorted client-side)

File: webui/react/src/hooks/useModal/Workspace/useModalWorkspaceCreate.tsx
Comment (mapmeld) [reply webui/react/src/hooks/useModal/Workspace/useModalWorkspaceCreate.tsx]: ok I removed the try/catch and `handleError` from the store, so `createWorkspace` can return a proper `Promise<Workspace>` and an error will travel up to the previous & visible handleError call.

**Comment by mapmeld** [comment]
> when workspace name already exists, it should show error notification. i think it's a regression

Latest commit should fix this by moving which try/catch covers this

File: webui/react/src/stores/workspaces.tsx
Comment (keita-determined) [inline webui/react/src/stores/workspaces.tsx:139]: this `await`  has no effect

**Review by keita-determined — APPROVED** [review]
LGTM for frontend part with one comment
should we wait for backend review too?

File: webui/react/src/stores/workspaces.tsx
Comment (mapmeld) [reply webui/react/src/stores/workspaces.tsx:139]: is it needed to wait for possible errors when deleting?

Comment (keita-determined) [reply webui/react/src/stores/workspaces.tsx:139]: Editor suggests it 
`updateWorkspaces()` is a normal func 

image

**Comment by ClaireNeveu** [comment]
Server-side changes here are very simple, you can merge without explicit approval from a member of the backend team.

File: webui/react/src/stores/workspaces.tsx
Comment (mapmeld) [reply webui/react/src/stores/workspaces.tsx:139]: oh, that's a different section of this file (inside of `useUpdateWorkspace`), which only updates the local store, so I just pushed a commit to remove async/await there.

Inside of `useDeleteWorkspace` I am going to keep the async/await, because it makes a remote call and if we wanted any return values or additional lines of code, it would need to be async/await.

Comment (ClaireNeveu) [reply webui/react/src/stores/workspaces.tsx:139]: You also don't need a try/catch block since `updateWorkspaces` never throws.

Comment (mapmeld) [reply webui/react/src/stores/workspaces.tsx:135]: ✅  changed Promise<void> to void, removed the try-catch,

**Pull request merged**
